tests/jwt_test.py

```diff
@@ -252,6 +260,156 @@ def test_jwt_with_sgx_key_filter(network, args):
+                {"table": "public:ccf.gov.jwt_public_signing_keys", "key": kid},
+            )
+            if cert_pem is None:
+                assert r.status_code == 400, r.status_code
+            else:
+                assert r.status_code == 200, r.status_code
+                # Note that /gov/read returns all data as JSON.
+                # Here, the stored data is a uint8 array, therefore it
+                # is returned as an array of integers.
+                cert_kv_der = bytes(r.body.json())
+                cert_kv_pem = infra.crypto.cert_der_to_pem(cert_kv_der)
+                assert infra.crypto.are_certs_equal(
+                    cert_pem, cert_kv_pem
+                ), "stored cert not equal to input cert"
+
+    def get_jwt_refresh_endpoint_metrics() -> dict:
+        with primary.client(
+            f"member{network.consortium.get_any_active_member().member_id}"
+        ) as c:
+            r = c.get("/gov/endpoint_metrics")
+            m = r.body.json()["metrics"]["jwt_keys/refresh"]["POST"]
+            assert m["errors"] == 0, m["errors"]  # not used in jwt refresh endpoint
+            m["successes"] = m["calls"] - m["failures"]
+            return m
+
+    LOG.info("Start OpenID endpoint server")
+    jwks = create_jwks(kid, cert_pem)
+    with OpenIDProviderServer(issuer_port, key_priv_pem, cert_pem, jwks) as server:
+        LOG.info("Add JWT issuer with auto-refresh")
+        with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as metadata_fp:
+            json.dump(
+                {"issuer": issuer, "auto_refresh": True, "ca_cert_name": ca_cert_name},
+                metadata_fp,
+            )
+            metadata_fp.flush()
+            network.consortium.set_jwt_issuer(primary, metadata_fp.name)
+
+        LOG.info("Wait for key refresh to happen")
+        # Note: refresh interval is set to 1s, see network args below.
+        time.sleep(2)
+
+        LOG.info("Check that keys got refreshed")
+        check_kv_jwt_key_matches(kid, cert_pem)
+
+        LOG.info("Check that JWT refresh endpoint has no failures")
+        m = get_jwt_refresh_endpoint_metrics()
+        assert m["failures"] == 0, m["failures"]
+        assert m["successes"] > 0, m["successes"]
+
+        LOG.info("Serve invalid JWKS")
+        server.jwks = {"foo": "bar"}
+
+        LOG.info("Wait for key refresh to happen")
+        time.sleep(2)
+
+        LOG.info("Check that JWT refresh endpoint has some failures")
+        m = get_jwt_refresh_endpoint_metrics()
+        assert m["failures"] > 0, m["failures"]
+
+    LOG.info("Restart OpenID endpoint server with new keys")
+    kid2 = "my_kid_2"
+    key2_priv_pem, _ = infra.crypto.generate_rsa_keypair(2048)
+    cert2_pem = infra.crypto.generate_cert(key2_priv_pem, cn=issuer_host)
+    jwks = create_jwks(kid2, cert2_pem)
+    with OpenIDProviderServer(issuer_port, key_priv_pem, cert_pem, jwks):
+        LOG.info("Wait for key refresh to happen")
+        time.sleep(2)
```

Comment:
We do not want one-shot waits for asynchronous operations they're brittle. This should instead be implemented as a loop on a concrete condition with an eventual timeout that's long enough to work even when the CI is a bit slow.

Reply:
Done